subsys/bluetooth/controller/ll_sw/ull_conn.c

```diff
@@ -0,0 +1,4730 @@
+				      offsetof(struct pdu_data, llctrl) + \
+				      sizeof(struct pdu_data_llctrl)) * \
+			       CONN_TX_CTRL_BUFFERS)
+
+static MFIFO_DEFINE(conn_tx, sizeof(struct lll_tx),
+		    CONFIG_BT_CTLR_TX_BUFFERS);
+
+static struct {
+	void *free;
+	u8_t pool[CONN_TX_BUF_SIZE * CONFIG_BT_CTLR_TX_BUFFERS];
+} mem_conn_tx;
+
+static struct {
+	void *free;
+	u8_t pool[CONN_TX_CTRL_BUF_SIZE * CONN_TX_CTRL_BUFFERS];
+} mem_conn_tx_ctrl;
+
+static struct {
+	void *free;
+	u8_t pool[sizeof(memq_link_t) *
+		  (CONFIG_BT_CTLR_TX_BUFFERS + CONN_TX_CTRL_BUFFERS)];
+} mem_link_tx;
+
+static u8_t data_chan_map[5] = {0xFF, 0xFF, 0xFF, 0xFF, 0x1F};
+static u8_t data_chan_count = 37U;
+
+#if defined(CONFIG_BT_CTLR_PHY)
+static u8_t default_phy_tx;
+static u8_t default_phy_rx;
+#endif /* CONFIG_BT_CTLR_PHY */
+
+static struct ll_conn conn_pool[CONFIG_BT_MAX_CONN];
+static struct ll_conn *conn_upd_curr;
+static void *conn_free;
+
+static struct device *entropy;
+
+struct ll_conn *ll_conn_acquire(void)
+{
+	return mem_acquire(&conn_free);
+}
+
+void ll_conn_release(struct ll_conn *conn)
+{
+	mem_release(conn, &conn_free);
+}
+
+u16_t ll_conn_handle_get(struct ll_conn *conn)
+{
+	return mem_index_get(conn, conn_pool, sizeof(struct ll_conn));
+}
+
+struct ll_conn *ll_conn_get(u16_t handle)
+{
+	return mem_get(conn_pool, sizeof(struct ll_conn), handle);
+}
+
+struct ll_conn *ll_connected_get(u16_t handle)
+{
+	struct ll_conn *conn;
+
+	if (handle >= CONFIG_BT_MAX_CONN) {
+		return NULL;
+	}
+
+	conn = ll_conn_get(handle);
+	if (conn->lll.handle != handle) {
+		return NULL;
+	}
+
+	return conn;
+}
+
+void *ll_tx_mem_acquire(void)
+{
+	return mem_acquire(&mem_conn_tx.free);
+}
+
+void ll_tx_mem_release(void *tx)
+{
+	mem_release(tx, &mem_conn_tx.free);
+}
+
+int ll_tx_mem_enqueue(u16_t handle, void *tx)
+{
+	struct lll_tx *lll_tx;
+	struct ll_conn *conn;
+	u8_t idx;
+
+	conn = ll_connected_get(handle);
+	if (!conn) {
+		return -EINVAL;
+	}
+
+	idx = MFIFO_ENQUEUE_GET(conn_tx, (void **) &lll_tx);
+	if (!lll_tx) {
+		return -ENOBUFS;
+	}
+
+	lll_tx->handle = handle;
+	lll_tx->node = tx;
+
+	MFIFO_ENQUEUE(conn_tx, idx);
+
+	return 0;
+}
+
+u8_t ll_conn_update(u16_t handle, u8_t cmd, u8_t status, u16_t interval_min,
+		    u16_t interval_max, u16_t latency, u16_t timeout)
+{
+	struct ll_conn *conn;
+
+	conn = ll_connected_get(handle);
+	if (!conn) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
+
+	if (!cmd) {
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+		if (!conn->llcp_conn_param.disabled &&
+		    (!conn->common.fex_valid ||
+		     (conn->llcp_features &
+		      BIT(BT_LE_FEAT_BIT_CONN_PARAM_REQ)))) {
+			cmd++;
+		} else if (conn->lll.role) {
+			return BT_HCI_ERR_UNSUPP_REMOTE_FEATURE;
+		}
+#else /* !CONFIG_BT_CTLR_CONN_PARAM_REQ */
+		if (conn->lll.role) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
+#endif /* !CONFIG_BT_CTLR_CONN_PARAM_REQ */
+	}
+
+	if (!cmd) {
+		if (conn->llcp_req != conn->llcp_ack) {
```

Comment:
The problem I see with these is that the register are typically loaded first, and then the comparison is executed:

```
                if (conn->llcp_req != conn->llcp_ack) {
   13916:       f890 30d0       ldrb.w  r3, [r0, #208]  ; 0xd0
   1391a:       f890 20d1       ldrb.w  r2, [r0, #209]  ; 0xd1
   1391e:       429a            cmp     r2, r3
```

Which means that if you get interrupted between `1391a` and `1391e` you will be comparing a stale value.

Reply:
I will add a commit to this PR to handle context-safety of the control procedures.